File: morph/importcol/entities.go
package importcol

import (
	"encoding/json"
	"strings"
)

type EntityKind string

const (
	EntityDistrict EntityKind = "district"
	EntityFacility EntityKind = "facility"
	EntityMember   EntityKind = "member"
	EntityEmployee EntityKind = "employee"
	EntityAsset    EntityKind = "asset"
	EntityActivity EntityKind = "activity"
	EntityContact  EntityKind = "contact"
	EntityUser     EntityKind = "user"
)

func ParseEntityKind(s string) (EntityKind, bool) {
	switch strings.ToLower(strings.TrimSpace(s)) {
	case "district", "districts":
		return EntityDistrict, true
	case "facility", "facilities", "school", "schools", "place", "places":
		return EntityFacility, true
	case "member", "members", "participant", "participants", "student", "students", "people-member", "people":
		return EntityMember, true
	case "employee", "employees", "staff", "people-employee":
		return EntityEmployee, true
	case "asset", "assets", "vehicle", "vehicles":
		return EntityAsset, true
	case "activity", "activities", "trip", "trips":
		return EntityActivity, true
	case "contact", "contacts":
		return EntityContact, true
	case "user", "users":
		return EntityUser, true
	default:
		return "", false
	}
}

func (k EntityKind) Label() string {
	switch k {
	case EntityDistrict:
		return "District"
	case EntityFacility:
		return "Places"
	case EntityMember:
		return "People"
	case EntityEmployee:
		return "People — employees"
	case EntityAsset:
		return "Assets"
	case EntityActivity:
		return "Activities"
	case EntityContact:
		return "Contact"
	case EntityUser:
		return "Morph User"
	default:
		return string(k)
	}
}

func AllKinds() []EntityKind {
	return []EntityKind{
		EntityMember, EntityAsset, EntityActivity,
	}
}

type EntitySpec struct {
	Kind             string   `json:"kind"`
	Label            string   `json:"label"`
	Description      string   `json:"description"`
	RequiredFields   []string `json:"required_fields"`
	OptionalFields   []string `json:"optional_fields"`
	TemplateHeaders  []string `json:"template_headers"`
	CSVTemplate      string   `json:"csv_template"`
	CSVExample       string   `json:"csv_example"`
	JSONExample      string   `json:"json_example"`
	Instructions     []string `json:"instructions"`
}

func SpecFor(kind EntityKind) EntitySpec {
	var optional, headers []string
	var csvEx, jsonEx, desc string
	switch kind {
	case EntityDistrict:
		optional = []string{"district_id", "district", "description"}
		headers = []string{"name", "district_id", "district", "description"}
		csvEx = "Central Region,DR-001,Central Region,Regional district for central facilities"
		jsonEx = `[{"name":"Central Region","district_id":"DR-001","district":"Central Region","description":"Regional district"}]`
		desc = "Import district records into MorphData District table."
	case EntityFacility:
		optional = []string{"facility_code", "district_id", "facility_type", "description", "location"}
		headers = []string{"name", "facility_code", "district_id", "facility_type", "description", "location"}
		csvEx = "Main Campus,FC-100,1,school,Primary campus building,"
		jsonEx = `[{"name":"Main Campus","facility_code":"FC-100","district_id":"1","facility_type":"school","description":"Primary campus"}]`
		desc = "Import places (sites) linked to districts. Columns: name, code (facility_code), type (facility_type)."
	case EntityMember:
		optional = []string{"first_name", "middle_name", "dob", "entry_date", "facility", "gender", "email", "participant_type", "description"}
		headers = []string{"last_name", "first_name", "middle_name", "dob", "entry_date", "facility", "gender", "email", "participant_type", "description"}
		csvEx = "Doe,Jane,,2005-03-15,2024-09-01,Main Campus,1,jane@example.com,student,Grade 5 participant"
		jsonEx = `[{"last_name":"Doe","first_name":"Jane","email":"jane@example.com","facility":"Main Campus","description":"Grade 5 participant"}]`
		desc = "Import people who are members/participants. The facility column is a place name or code."
	case EntityEmployee:
		optional = []string{"first_name", "middle_name", "staff_guid", "email", "phone_number", "facility_id", "employ_type", "description"}
		headers = []string{"last_name", "first_name", "middle_name", "staff_guid", "email", "phone_number", "facility_id", "employ_type", "description"}
		csvEx = "Smith,John,,EMP-42,john@example.com,555-0100,1,full-time,Transport coordinator"
		jsonEx = `[{"last_name":"Smith","first_name":"John","email":"john@example.com","facility_id":"1","description":"Transport coordinator"}]`
		desc = "Import people who are employees. facility_id is the place record id."
	case EntityAsset:
		optional = []string{"description", "AssetType", "ContractorID"}
		headers = []string{"asset_tag", "description", "AssetType", "ContractorID"}
		csvEx = "BUS-12,School bus type C,bus,"
		jsonEx = `[{"asset_tag":"BUS-12","description":"School bus type C","AssetType":"bus"}]`
		desc = "Import assets (vehicles, equipment)."
	case EntityActivity:
		optional = []string{"ActivityType", "start_date", "end_date", "location", "GUID", "description"}
		headers = []string{"Name", "ActivityType", "start_date", "end_date", "location", "GUID", "description"}
		csvEx = "Field Trip 2026,excursion,2026-05-01,2026-05-01,Museum,ACT-001,Annual museum visit"
		jsonEx = `[{"Name":"Field Trip 2026","ActivityType":"excursion","start_date":"2026-05-01","description":"Annual museum visit"}]`
		desc = "Import activities/trips."
	case EntityContact:
		optional = []string{"FirstName", "Email", "Phone", "Mobile", "description"}
		headers = []string{"LastName", "FirstName", "Email", "Phone", "Mobile", "description"}
		csvEx = "Doe,Jane,jane@example.com,555-0101,,Billing contact"
		jsonEx = `[{"LastName":"Doe","FirstName":"Jane","Email":"jane@example.com","description":"Billing contact"}]`
		desc = "Import contacts."
	case EntityUser:
		optional = []string{"email", "phone", "administrator"}
		headers = []string{"login_id", "first_name", "last_name", "email", "phone", "administrator"}
		csvEx = "jdoe,Jane,Doe,jane@example.com,555-0100,false"
		jsonEx = `[{"login_id":"jdoe","first_name":"Jane","last_name":"Doe","email":"jane@example.com","administrator":"false"}]`
		desc = "Import Morph User profiles."
	}
	headerLine := strings.Join(headers, ",")
	return EntitySpec{
		Kind:            string(kind),
		Label:           kind.Label(),
		Description:     desc,
		RequiredFields:  []string{},
		OptionalFields:  optional,
		TemplateHeaders: headers,
		CSVTemplate:     headerLine + "\n",
		CSVExample:      headerLine + "\n" + csvEx + "\n",
		JSONExample:     jsonEx + "\n",
		Instructions: []string{
			"Download the CSV or JSON example and add your rows.",
			"Using the template: template columns import into MorphData.",
			"Validate your file, then start the background import.",
		},
	}
}

func AllSpecs() []EntitySpec {
	out := make([]EntitySpec, 0, len(AllKinds()))
	for _, k := range AllKinds() {
		out = append(out, SpecFor(k))
	}
	return out
}

// MustEncode is a tiny helper for tests/debug.
func MustEncode(v any) string {
	b, _ := json.Marshal(v)
	return string(b)
}

File: UsersPanel/backend/src/importcol/entities.rs
use serde::{Deserialize, Serialize};

/// MorphData entity types supported by Data Collector.
#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum EntityKind {
    District,
    Facility,
    Member,
    Employee,
    Asset,
    Activity,
    Contact,
    User,
}

impl EntityKind {
    pub fn from_str(s: &str) -> Option<Self> {
        match s.trim().to_lowercase().as_str() {
            "district" | "districts" => Some(Self::District),
            "facility" | "facilities" | "school" | "schools" => Some(Self::Facility),
            "member" | "members" | "participant" | "participants" | "student" | "students" => {
                Some(Self::Member)
            }
            "employee" | "employees" | "staff" => Some(Self::Employee),
            "asset" | "assets" | "vehicle" | "vehicles" => Some(Self::Asset),
            "activity" | "activities" | "trip" | "trips" => Some(Self::Activity),
            "contact" | "contacts" => Some(Self::Contact),
            "user" | "users" => Some(Self::User),
            _ => None,
        }
    }

    pub fn as_str(self) -> &'static str {
        match self {
            Self::District => "district",
            Self::Facility => "facility",
            Self::Member => "member",
            Self::Employee => "employee",
            Self::Asset => "asset",
            Self::Activity => "activity",
            Self::Contact => "contact",
            Self::User => "user",
        }
    }

    pub fn label(self) -> &'static str {
        match self {
            Self::District => "District",
            Self::Facility => "Facility",
            Self::Member => "Participant (Member)",
            Self::Employee => "Employee",
            Self::Asset => "Asset",
            Self::Activity => "Activity",
            Self::Contact => "Contact",
            Self::User => "Morph User",
        }
    }

    pub fn all() -> &'static [EntityKind] {
        &[
            Self::District,
            Self::Facility,
            Self::Member,
            Self::Employee,
            Self::Asset,
            Self::Activity,
            Self::Contact,
            Self::User,
        ]
    }
}

#[derive(Debug, Clone, Serialize)]
pub struct EntitySpec {
    pub kind: EntityKind,
    pub label: String,
    pub description: String,
    pub required_fields: Vec<String>,
    pub optional_fields: Vec<String>,
    pub template_headers: Vec<String>,
    pub csv_template: String,
    pub csv_example: String,
    pub json_example: String,
    pub instructions: Vec<String>,
}

pub fn spec_for(kind: EntityKind) -> EntitySpec {
    let (_required, optional, headers, csv_ex, json_ex) = match kind {
        EntityKind::District => (
            vec!["name"],
            vec!["district_id", "district", "description"],
            vec!["name", "district_id", "district", "description"],
            "Central Region,DR-001,Central Region,Regional district for central facilities",
            r#"[{"name":"Central Region","district_id":"DR-001","district":"Central Region","description":"Regional district"}]"#,
        ),
        EntityKind::Facility => (
            vec!["name"],
            vec![
                "facility_code",
                "district_id",
                "facility_type",
                "description",
                "location",
            ],
            vec![
                "name",
                "facility_code",
                "district_id",
                "facility_type",
                "description",
                "location",
            ],
            "Main Campus,FC-100,1,school,Primary campus building,\"{\"\"city\"\":\"\"Springfield\"\"}\"",
            r#"[{"name":"Main Campus","facility_code":"FC-100","district_id":"1","facility_type":"school","description":"Primary campus"}]"#,
        ),
        EntityKind::Member => (
            vec!["last_name"],
            vec![
                "first_name",
                "middle_name",
                "dob",
                "entry_date",
                "facility",
                "gender",
                "email",
                "participant_type",
                "description",
            ],
            vec![
                "last_name",
                "first_name",
                "middle_name",
                "dob",
                "entry_date",
                "facility",
                "gender",
                "email",
                "participant_type",
                "description",
            ],
            "Doe,Jane,,2005-03-15,2024-09-01,Main Campus,1,jane@example.com,student,Grade 5 participant",
            r#"[{"last_name":"Doe","first_name":"Jane","email":"jane@example.com","facility":"Main Campus","description":"Grade 5 participant"}]"#,
        ),
        EntityKind::Employee => (
            vec!["last_name"],
            vec![
                "first_name",
                "middle_name",
                "staff_guid",
                "email",
                "phone_number",
                "facility_id",
                "employ_type",
                "description",
            ],
            vec![
                "last_name",
                "first_name",
                "middle_name",
                "staff_guid",
                "email",
                "phone_number",
                "facility_id",
                "employ_type",
                "description",
            ],
            "Smith,John,,EMP-42,john@example.com,555-0100,1,full-time,Transport coordinator",
            r#"[{"last_name":"Smith","first_name":"John","email":"john@example.com","facility_id":"1","description":"Transport coordinator"}]"#,
        ),
        EntityKind::Asset => (
            vec!["asset_tag"],
            vec!["description", "AssetType", "ContractorID"],
            vec!["asset_tag", "description", "AssetType", "ContractorID"],
            "BUS-12,School bus type C,bus,",
            r#"[{"asset_tag":"BUS-12","description":"School bus type C","AssetType":"bus"}]"#,
        ),
        EntityKind::Activity => (
            vec!["Name"],
            vec![
                "ActivityType",
                "start_date",
                "end_date",
                "location",
                "GUID",
                "description",
            ],
            vec![
                "Name",
                "ActivityType",
                "start_date",
                "end_date",
                "location",
                "GUID",
                "description",
            ],
            "Field Trip 2026,excursion,2026-05-01,2026-05-01,Museum,ACT-001,Annual museum visit",
            r#"[{"Name":"Field Trip 2026","ActivityType":"excursion","start_date":"2026-05-01","description":"Annual museum visit"}]"#,
        ),
        EntityKind::Contact => (
            vec!["LastName"],
            vec!["FirstName", "Email", "Phone", "Mobile", "description"],
            vec![
                "LastName",
                "FirstName",
                "Email",
                "Phone",
                "Mobile",
                "description",
            ],
            "Doe,Jane,jane@example.com,555-0101,,Billing contact",
            r#"[{"LastName":"Doe","FirstName":"Jane","Email":"jane@example.com","description":"Billing contact"}]"#,
        ),
        EntityKind::User => (
            vec!["login_id", "first_name", "last_name"],
            vec!["email", "phone", "administrator"],
            vec![
                "login_id",
                "first_name",
                "last_name",
                "email",
                "phone",
                "administrator",
            ],
            "jdoe,Jane,Doe,jane@example.com,555-0100,false",
            r#"[{"login_id":"jdoe","first_name":"Jane","last_name":"Doe","email":"jane@example.com","administrator":"false"}]"#,
        ),
    };

    let mut optional_fields: Vec<String> = optional.into_iter().map(String::from).collect();
    if !optional_fields
        .iter()
        .any(|field| field.eq_ignore_ascii_case("title"))
    {
        optional_fields.push("title".to_string());
    }
    let required_fields: Vec<String> = Vec::new();

    let header_line = headers.join(",");
    let csv_template = format!("{header_line}\n");
    let csv_example = format!("{header_line}\n{csv_ex}\n");
    let json_example = format!("{json_ex}\n");

    let description = match kind {
        EntityKind::District => "Import district records into MorphData District table.",
        EntityKind::Facility => "Import facilities (schools/sites) linked to districts.",
        EntityKind::Member => "Import participants (member table).",
        EntityKind::Employee => "Import employees/staff records.",
        EntityKind::Asset => "Import assets (vehicles, equipment).",
        EntityKind::Activity => "Import activities/trips.",
        EntityKind::Contact => "Import contacts.",
        EntityKind::User => "Import Morph User profiles.",
    };

    let instructions = vec![
        "Download the CSV or JSON example and add your rows.".to_string(),
        "Using the template: template columns import into MorphData; any extra columns become JSON Details.".to_string(),
        "Without the template: matching column names import into MorphData; everything else becomes JSON Details.".to_string(),
        "Validate your file, then start the background import.".to_string(),
    ];

    EntitySpec {
        kind,
        label: kind.label().to_string(),
        description: description.to_string(),
        required_fields,
        optional_fields,
        template_headers: headers.into_iter().map(String::from).collect(),
        csv_template,
        csv_example,
        json_example,
        instructions,
    }
}

pub fn all_specs() -> Vec<EntitySpec> {
    EntityKind::all()
        .iter()
        .copied()
        .map(spec_for)
        .collect()
}
